src/types/redis_hyperloglog.cc

```diff
@@ -0,0 +1,359 @@
+    auto s = storage_->Get(rocksdb::ReadOptions(), sub_key, &old_count);
+    if (!s.ok() && !s.IsNotFound()) return s;
+    if (static_cast<unsigned long>(count) > std::stoul(old_count)) {
+      batch->Put(sub_key, std::to_string(count));
+      *ret = 1;
+    }
+  }
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status Hyperloglog::Count(const Slice &user_key, int *ret) {
+  *ret = 0;
+  std::vector<uint8_t> counts(HLL_REGISTERS);
+  auto s = getRegisters(user_key, &counts);
+  if (!s.ok()) return s;
+  *ret = hllCount(counts);
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status Hyperloglog::Merge(const std::vector<Slice> &user_keys) {
+  std::vector<uint8_t> max(HLL_REGISTERS);
+  for (const auto &user_key : user_keys) {
+    std::vector<uint8_t> counts(HLL_REGISTERS);
+    auto s = getRegisters(user_key, &counts);
+    if (!s.ok()) return s;
+    hllMerge(&max[0], counts);
+  }
+
+  std::string ns_key = AppendNamespacePrefix(user_keys[0]);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  HyperloglogMetadata metadata;
+  rocksdb::Status s = GetMetadata(ns_key, &metadata);
+  if (!s.ok() && !s.IsNotFound()) return s;
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisHyperloglog);
+  batch->PutLogData(log_data.Encode());
+  if (s.IsNotFound()) {
+    std::string bytes;
+    metadata.size = HLL_REGISTERS;
+    metadata.Encode(&bytes);
+    batch->Put(metadata_cf_handle_, ns_key, bytes);
+  }
+
+  for (auto i = 0; i < HLL_REGISTERS; i++) {
+    std::string sub_key = InternalKey(ns_key, std::to_string(i), metadata.version, storage_->IsSlotIdEncoded()).Encode();
+
+    batch->Put(sub_key, std::to_string(max[i]));
+  }
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+/* Store the value of the register at position 'regnum' into variable 'target'.
+ * 'p' is an array of unsigned bytes. */
+#define HLL_DENSE_GET_REGISTER(target,p,regnum) do { \
+    uint8_t *_p = (uint8_t*) p; \
+    unsigned long _byte = regnum*HLL_BITS/8; \
+    unsigned long _fb = regnum*HLL_BITS&7; \
+    unsigned long _fb8 = 8 - _fb; \
+    unsigned long b0 = _p[_byte]; \
+    unsigned long b1 = _p[_byte+1]; \
+    target = ((b0 >> _fb) | (b1 << _fb8)) & HLL_REGISTER_MAX; \
+} while(0)
+
+/* ========================= HyperLogLog algorithm  ========================= */
+
+/* Our hash function is MurmurHash2, 64 bit version.
+ * It was modified for Redis in order to provide the same result in
+ * big and little endian archs (endian neutral). */
+uint64_t MurmurHash64A (const void * key, int len, unsigned int seed) {
+    const uint64_t m = 0xc6a4a7935bd1e995;
+    const int r = 47;
+    uint64_t h = seed ^ (len * m);
+    const uint8_t *data = (const uint8_t *)key;
+    const uint8_t *end = data + (len-(len&7));
+
+    while(data != end) {
+        uint64_t k;
+
+#if (BYTE_ORDER == LITTLE_ENDIAN)
+    #ifdef USE_ALIGNED_ACCESS
```

Comment:
Would you mind extract `USE_ALIGNED_ACCESS` from `redis_bitmap.cc`?

Reply:
sorry, I do not understand ''extract" clearly, do you mean to add a cmake compile option for USE_ALIGNED_ACCESS?

Reply:
Currently, `USE_ALIGNED_ACCESS` is defined here: https://github.com/apache/kvrocks/blob/83b3adfdeabec572ec00bb15e93eb6f4000dfb59/src/types/redis_bitmap.h#L32

It would be only used in `redis_bitmap.cc`. You can extract it to a common header or define it in cmake

Reply:
ok, got it.

Reply:
Had moved to 'storage.h'.